src/mallocTracer.cpp

```diff
@@ -0,0 +1,248 @@
+    void* addr = dlsym(RTLD_NEXT, #sym); \
+    addr != NULL ? (sym##_t)addr : sym;  \
+})
+
+typedef void* (*malloc_t)(size_t);
+static malloc_t _orig_malloc = NULL;
+
+typedef void* (*calloc_t)(size_t, size_t);
+static calloc_t _orig_calloc = NULL;
+
+typedef void* (*realloc_t)(void*, size_t);
+static realloc_t _orig_realloc = NULL;
+
+typedef void (*free_t)(void*);
+static free_t _orig_free = NULL;
+
+static void* malloc_hook(size_t size) {
+    void* ret = _orig_malloc(size);
+    if (likely(ret && size)) {
+        MallocTracer::recordMalloc(ret, size);
+    }
+    return ret;
+}
+
+#ifdef __linux__
+extern "C" WEAK DLLEXPORT void* malloc(size_t size) {
+    if (unlikely(!_orig_malloc)) {
+        return NULL;
+    }
+
+    if (likely(MallocTracer::initialized())) {
+        return malloc_hook(size);
+    }
+    return _orig_malloc(size);
+}
+#endif
+
+static void* calloc_hook(size_t num, size_t size) {
+    void* ret = _orig_calloc(num, size);
+    if (likely(ret && num && size)) {
+        MallocTracer::recordMalloc(ret, num * size);
+    }
+    return ret;
+}
+
+#ifdef __linux__
+extern "C" WEAK DLLEXPORT void* calloc(size_t num, size_t size) {
+    if (unlikely(!_orig_calloc)) {
+        return NULL;
+    }
+
+    if (likely(MallocTracer::initialized())) {
+        return calloc_hook(num, size);
+    }
+    return _orig_calloc(num, size);
+}
+#endif
+
+static void* realloc_hook(void* addr, size_t size) {
+    void* ret = _orig_realloc(addr, size);
+    if (likely(ret && addr)) {
+        MallocTracer::recordFree(addr);
+    }
+
+    if (likely(ret && size)) {
+        MallocTracer::recordMalloc(ret, size);
+    }
+    return ret;
+}
+
+#ifdef __linux__
+extern "C" WEAK DLLEXPORT void* realloc(void* addr, size_t size) {
+    if (unlikely(!_orig_realloc)) {
+        return NULL;
+    }
+
+    if (likely(MallocTracer::initialized())) {
+        return realloc_hook(addr, size);
+    }
+    return _orig_realloc(addr, size);
+}
+#endif
+
+static void free_hook(void* addr) {
+    _orig_free(addr);
+    if (addr) {
+        MallocTracer::recordFree(addr);
+    }
+}
+
+#ifdef __linux__
+extern "C" WEAK DLLEXPORT void free(void* addr) {
+    if (unlikely(!_orig_free)) {
+        return;
+    }
+
+    if (likely(MallocTracer::initialized())) {
+        return free_hook(addr);
+    }
+    return _orig_free(addr);
+}
+#endif
+
+u64 MallocTracer::_interval;
+volatile u64 MallocTracer::_allocated_bytes;
+
+Mutex MallocTracer::_patch_lock;
+int MallocTracer::_patched_libs = 0;
+bool MallocTracer::_initialized = false;
+
+// Only to be used from getOrigAddresses
+void* safeCalloc(size_t num, size_t size) {
+    void* ret = OS::safeAlloc(num * size);
+    if (likely(ret)) {
+        memset(ret, 0, num * size);
+    }
+    return ret;
+}
+
+__attribute__((constructor)) static void getOrigAddresses() {
+    // Set malloc and calloc which may be called from libc during getOrigAddresses.
+    _orig_malloc = OS::safeAlloc;
```

Comment:
So is it really necessary? I remember you were saying allocation in dlsym only happens on error path, and returning NULL does not break anything.

Reply:
For a specific libc version yes. Removing to see test results.